clippy_lints/src/clamping_without_clamp.rs

```diff
@@ -0,0 +1,766 @@
+    Mutability, Pat, PatKind, QPath, Stmt, StmtKind,
+};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::Ty;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::Span;
+use std::ops::Deref;
+
+use clippy_utils::{diagnostics::span_lint_and_sugg, get_trait_def_id, paths, sugg::Sugg, ty::implements_trait};
+use rustc_errors::Applicability;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Identifies good opportunities for a clamp function from std or core, and suggests using it.
+    ///
+    /// ### Why is this bad?
+    /// clamp is much shorter, easier to read, and doesn't use any control flow.
+    ///
+    /// ### Known issue(s)
+    /// `clamp` functions will panic if `max < min`, `max.is_nan()`, or `min.is_nan()`.
+    /// Some may consider panicking in these situations to be desirable, but it also may
+    /// introduce panicking where there wasn't any before.
+    ///
+    /// ### Examples
+    /// ```rust
+    /// # let (input, min, max) = (0, -2, 1);
+    /// if input > max {
+    ///     max
+    /// }
+    /// else if input < min {
+    ///     min
+    /// } else {
+    ///     input
+    /// }
+    /// # ;
+    /// ```
+    ///
+    /// ```rust
+    /// # let (input, min, max) = (0, -2, 1);
+    /// input.max(min).min(max)
+    /// # ;
+    /// ```
+    ///
+    /// ```rust
+    /// # let (input, min, max) = (0, -2, 1);
+    /// match input {
+    ///     input if input > max => max,
+    ///     input if input < min => min,
+    ///     input => input,
+    /// }
+    /// # ;
+    /// ```
+    ///
+    /// ```rust
+    /// # let (input, min, max) = (0, -2, 1);
+    /// let mut x = input;
+    /// if x < min { x = min; }
+    /// if x > max { x = max; }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// # let (input, min, max) = (0, -2, 1);
+    /// input.clamp(min, max)
+    /// # ;
+    /// ```
+    #[clippy::version = "1.65.0"]
+    pub CLAMPING_WITHOUT_CLAMP,
+    complexity,
+    "using a clamp pattern instead of the clamp function"
+}
+declare_lint_pass!(ClampingWithoutClamp => [CLAMPING_WITHOUT_CLAMP]);
+
+type ExprRef<'tcx> = &'tcx Expr<'tcx>;
+
+#[derive(Debug)]
+struct InputMinMax<'tcx> {
+    input: ExprRef<'tcx>,
+    min: ExprRef<'tcx>,
+    max: ExprRef<'tcx>,
+    add_semicolon: bool,
+    span: Span,
+}
+
+impl<'tcx> LateLintPass<'tcx> for ClampingWithoutClamp {
+    fn check_expr(&mut self, ctx: &LateContext<'tcx>, expr: ExprRef<'tcx>) {
+        if !expr.span.from_expansion() {
+            check_anti_patterns(ctx, expr);
+        }
+    }
+}
+
+fn check_anti_patterns<'tcx>(cx: &LateContext<'tcx>, expr: ExprRef<'tcx>) {
+    let suggestion_iter = is_if_else_pattern(cx, expr)
+        .or_else(|| is_max_min_pattern(cx, expr))
+        .or_else(|| is_match_pattern(cx, expr))
+        .into_iter()
+        .chain(is_two_if_pattern(cx, expr).into_iter());
```

Comment:
There's a good overlap in behaviour between the `is_if_else_pattern`, `is_match_pattern` and `is_two_if_pattern` cases:

```rust
if A1 OP_A A2 {
    A3
} else if B1 OP_B B2 {
    B3
} else {
    input
}

match .. {
    input if A1 OP_A A2 => A3,
    input if B1 OP_B B2 => B3,
    input => input,
}

let mut x = input;
if A1 OP_A A2 { x = A3; }
if B1 OP_B B2 { x = B3; }
```

It would be good to check these common parts in a function that all 3 call

Reply:
This was a great review comment, thank you! Implemented